### PR TITLE
Compact groups and logging optimization

### PR DESCRIPTION
**Pull request by Apollon77**
Hey,

so here wo go :-)
Many topics wnt into this PR (sorry for that but happened "on the way"). 

**Best Review on a Commit base. I tried to structure them a bit.**

Working for me, also with tested edge cases. The solution is not that "nice", but the alternative would had been a near to complete modularization and rewrite of the "controller.js" which is kind of stable stuff, so I decided against it for now.

Looking forwards to your comments.

I put it here as draft PR for now ecause I still need to get the testing green again and do not know if I manage that today (starting tomorrow on vacation for 1 week). But I will finalize when back and then we would be to finalize 2.0.0!

### CODE REVIEW

**Comment by Apollon77** [comment]
testing is green now :-)

We need to do one decision: https://github.com/ioBroker/ioBroker.js-controller/pull/428/files#diff-7a9076d6d94e62c13d641aa71f19ae8eR2078 ... I, for now, implemented it as it was till now: As soon as Adapter supports compact it is automatically enabled for this instance and runs in compact mode.
Do we want to have it that way? Or should we disable by default?
For future the idea would be to add a flag in the repo for "compactModeKnownStable" for each adapter and this could be used for this decision, but we do not have this now (should we add before 2.0)?

**Comment by GermanBluefox** [comment]
>should we add before 2.0

Flag is OK, but don't add it in 1.5.x

**Comment by Apollon77** [comment]
Before 2.0 was meaned as „add it to repo file before we release 2.0“ ...

**Comment by Apollon77** [comment]
... the more interesting question is how we do the process to allow someone set this flag to true ;-)

Should we enhance statistics to also send compact mode active Info beside Version and also show in in adapter list (x% run in compact Mode)

And which adapters we should set initially? But we should discuss this finally in the separate issue ;-)

**Comment by GermanBluefox** [comment]
Of course I am not able to check all 1000 lines of changes, but it looks OK.

**Comment by Apollon77** [comment]
As said check the three commits for its own to get a better scope. This should allow a review. Also changes to adapter.js is mainly logging and smaller stuff. 

I will rebase the PR on latest master when I‘m back to solve merge conflicts before and also test it before i really merge it. I already did one rebase and it only will pot. conflict with your changes in adapter.js from the alias feature. All other files should be not changed by anyone else (unless someone did in last 6 days ;-) )

**Comment by GermanBluefox** [comment]
> As said check the three commits for its own to get a better scope. This should allow a review. Also changes to adapter.js is mainly logging and smaller stuff.
> 
> I will rebase the PR on latest master when I‘m back to solve merge conflicts before and also test it before i really merge it. I already did one rebase and it only will pot. conflict with your changes in adapter.js from the alias feature. All other files should be not changed by anyone else (unless someone did in last 6 days ;-) )

I did https://github.com/ioBroker/ioBroker.js-controller/tree/multihost-upload

**Comment by Apollon77** [comment]
Kk. I will merge that ;-)

**Comment by Apollon77** [comment]
Ps: can you remove the numbers in the alias log messages again because then I would not have that many conflicts on merge ;-)

Merge will be fun but possible.

File: lib/setup/setupInstall.js [hunk 1]
```diff
@@ -829,8 +841,7 @@ function Install(options) {
                                         unit: '% of one core'
                                     },
                                     native: {}
-                                },
-                                {
+                                },                                {
```
Comment (AlCalzone) [inline lib/setup/setupInstall.js]: Formatting broke?

File: lib/states/statesInRedis.js [hunk 2]
```diff
@@ -129,13 +129,13 @@ class StateRedis {
             this.client.disconnect();
             ignoreErrors = true;
 
-            this.log.silly('Initiate Fallback to SocketIo States');
+            this.log.silly(this.namespace + ' Initiate Fallback to SocketIo States');
```
Comment (AlCalzone) [inline lib/states/statesInRedis.js]: isn't that spelled like `socket.io`, not `SocketIo`?

File: lib/cli/cliCompact.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    /**
+     * Set compact mode for this host
+     * @param targetState target state to set for host
+     */
+    setCompactModeHost(targetState) {
+        const { callback, dbConnect } = this.options;
+
+        dbConnect((_objects, _states, isOffline) => {
+            const compactMode = !!(this.config && this.config.system && this.config.system.compact);
```
Comment (AlCalzone) [inline lib/cli/cliCompact.js]: I think `currentState` would be more logical. I had to think about what line 80 does.

File: lib/cli/cliCompact.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            this.config.system.compact = targetState;
+            const configFileName = tools.getConfigFileName();
+            fs.writeFileSync(configFileName, JSON.stringify(this.config, null, 4), 'utf8');
+
+            console.log();
+            console.log('Compact mode for this host changed to ' + (targetState ? 'enabled' : 'disabled'));
+
+            if (!isOffline) {
+                console.log();
+                console.log('Please stop ioBroker for this change to take effect!')
```
Comment (AlCalzone) [inline lib/cli/cliCompact.js]: Wouldn't restarting instead of only stopping be more logical from a UX perspective?

File: lib/cli/cliCompact.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * Output the compact mode status for a defined adapter instance
+     * @param instance instance to otput status for
+     */
+    statusCompactModeInstance(instance) {
+        const { callback, dbConnect } = this.options;
+
+        dbConnect((objects, _states, isOffline, _objectDbType, config) => {
+            if (!this.config.system || !this.config.system.compact) {
+                console.log('Compact mode is disabled for this host. This setting will have no effect!');
```
Comment (AlCalzone) [inline lib/cli/cliCompact.js]: Which setting do you mean? This only outputs the current status.
How about "This setting currently has no effect!" ?

File: lib/cli/cliCompact.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            });
+        });
+    }
+
+    setCompactModeInstance(instance, targetState, groupId) {
+        const { callback, dbConnect } = this.options;
+
+        dbConnect((objects, _states, isOffline, _objectDbType, config) => {
+            if (!this.config.system || !this.config.system.compact) {
+                console.log('Compact mode is disabled on this host. This setting will have no effect!');
```
Comment (AlCalzone) [inline lib/cli/cliCompact.js]: Here it makes sense

File: lib/cli/cliCompact.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+                        if (newRunAsCompactMode !== undefined) {
+                            obj.common.runAsCompactMode = newRunAsCompactMode;
+                        }
+
+                        obj.from = 'system.host.' + tools.getHostName() + '.cli';
+                        obj.ts = new Date().getTime();
+                        objects.setObject('system.adapter.' + instance, obj, (err) => {
+                            if (err) console.log(err);
+                            console.log('Instance settings for "' + instance + '" are changed.');
+                            return void callback();
```
Comment (AlCalzone) [inline lib/cli/cliCompact.js]: The `void` is not really necessary here since it is inside another callback. But doesn't hurt.

File: main.js [hunk 8]
```diff
@@ -200,24 +206,24 @@ function logRedirect(isActive, id, reason) {
...
+                logger.error(hostLogPrefix + ' change event with no ID: ' + JSON.stringify(state));
                 return;
             }
             // If some log transporter activated or deactivated
             if (id.match(/.logging$/)) {
                 logRedirect(state ? state.val : false, id.substring(0, id.length - '.logging'.length), id);
             } else
             // If this is messagebox
-            if (id === 'messagebox.system.host.' + hostname) {
+            if (!compactGroupController && id === 'messagebox.' + hostObjectPrefix) {
```
Comment (AlCalzone) [inline main.js]: I'm not very deep into the inner workings of this. Why is this not allowed for compactGroupController ?

File: main.js [hunk 9]
```diff
@@ -318,9 +324,11 @@ function createStates(onConnect) {
         },
         connected: (handler) => {
             states = handler;
-            if (states.clearAllLogs)     states.clearAllLogs();
-            if (states.clearAllMessages) states.clearAllMessages();
-            deleteAllZipPackages();
+            if (!compactGroupController) {
```
Comment (AlCalzone) [inline main.js]: Here it is more clear, but these fine things deserve a comment IMO. Something like "logs are handled by the main controller".

File: main.js [hunk 10]
```diff
@@ -382,70 +390,108 @@ function createObjects(onConnect) {
...
                     // if during stopping the DB has connection again
                     if (connected && !isStopping) {
+                        logger.warn(hostLogPrefix + ' Slave controller has connection again ... restarting');
                         getInstances();
                         startAliveInterval();
                         initMessageQueue();
                     }
                 });
-            }, config.objects.connectTimeout || 2000);
+            }, (config.objects.connectTimeout || 2000) + (!compactGroupController ? 10 : 0));
```
Comment (AlCalzone) [inline main.js]: Hä?

File: main.js [hunk 11]
```diff
@@ -382,70 +390,108 @@ function createObjects(onConnect) {
...
-                    if (_ipArr.indexOf(obj.common.host) !== -1 || obj.common.host === hostname) {
-                        procs[id] = {config: obj};
-                        if (procs[id].config.common.enabled && (!procs[id].config.common.webExtension || !procs[id].config.native.webInstance)) {
-                            startInstance(id);
-                        }
+                    if (checkAndAddInstance(obj, _ipArr) && procs[id].config.common.enabled && (!procs[id].config.common.webExtension || !procs[id].config.native.webInstance)) {
+                            // We should give is a slight delay to allow an pot. former existing process on other host to exit
+                            if (compactGroupController) {
+                                procs[id].restartTimer = setTimeout(_id => startInstance(_id), 2500, id);
+                            }
```
Comment (AlCalzone) [inline main.js]: Formatting is weird here.

File: main.js [hunk 12]
```diff
@@ -382,70 +390,108 @@ function createObjects(onConnect) {
...
-
                 } else if (obj && obj.common) {
                     const _ipArr = getIPs();
                     // new adapter
-                    if (_ipArr.indexOf(obj.common.host) !== -1 || obj.common.host === hostname) {
-                        procs[id] = {config: obj};
-                        if (procs[id].config.common.enabled && (!procs[id].config.common.webExtension || !procs[id].config.native.webInstance)) {
-                            startInstance(id);
-                        }
+                    if (checkAndAddInstance(obj, _ipArr) && procs[id].config.common.enabled && (!procs[id].config.common.webExtension || !procs[id].config.native.webInstance)) {
```
Comment (AlCalzone) [inline main.js]: This check also deserves a comment. I have no clue why all these conditions are necessary.

File: main.js [hunk 13]
```diff
@@ -456,14 +502,21 @@ function startAliveInterval() {
     config.system = config.system || {};
     config.system.statisticsInterval = parseInt(config.system.statisticsInterval, 10) || 15000;
     config.system.checkDiskInterval  = (config.system.checkDiskInterval !== 0) ? parseInt(config.system.checkDiskInterval, 10) || 300000 : 0;
+    if (!compactGroupController) {
```
Comment (AlCalzone) [inline main.js]: Why do we need this state? --> Comment ;)

File: main.js [hunk 14]
```diff
@@ -541,6 +594,27 @@ function reportStatus() {
...
+           }
+           else {
+               realProcesses++;
+           }
+
+       }
+    });
+    states.setState(id + '.instancesAsProcess',   {val: realProcesses, ack: true, from: id});
+    states.setState(id + '.instancesAsCompact',   {val: compactProcesses, ack: true, from: id});
+
```
Comment (AlCalzone) [inline main.js]: Not too sure here... Is this supposed to be true?
`compactGroupProcesses = compactProcesses + realProcesses`
Because it might be that the else branch of line 601 is taken and then there would be a different count.

File: main.js [hunk 15]
```diff
@@ -541,6 +594,27 @@ function reportStatus() {
...
+    if (compactGroupController && started && compactProcesses === 0 && realProcesses === 0) {
+        logger.info('Compactgroup controller ' + compactGroup + ' does not own any processes, stop');
```
Comment (AlCalzone) [inline main.js]: `Compact group controller ...`

File: main.js [hunk 16]
```diff
@@ -668,6 +742,10 @@ function delObjects(objs, callback) {
  * @return none
  */
 function checkHost(callback) {
+    if (compactGroupController) {
```
Comment (AlCalzone) [inline main.js]: ?!?
Comment pls

File: main.js [hunk 17]
```diff
@@ -884,59 +961,76 @@ function extendObjects(tasks, callback) {
...
-                    title:      process.title,
-                    versions:   process.versions,
-                    env:        process.env
+        let newObj;
+        if (compactGroupController) {
+            newObj = {
+                _id: id,
+                type: 'device',
+                common: {
+                    name: hostname + '.compactgroup' + compactGroup,
```
Comment (AlCalzone) [inline main.js]: We need consistency. Sometimes you write `compactgroup`, sometimes `compactGroup`.

File: main.js [hunk 18]
```diff
@@ -884,59 +961,76 @@ function extendObjects(tasks, callback) {
...
+                    cmd: process.argv[0] + ' ' + (process.execArgv.join(' ') + ' ').replace(/--inspect-brk=\d+ /, '') + process.argv.slice(1).join(' '),
```
Comment (AlCalzone) [inline main.js]: Isn't this basically just the command line with `--inspect-brk` filtered out? Looks pretty complicated.

File: main.js [hunk 19]
```diff
@@ -1252,25 +1411,32 @@ function setMeta() {
     }
 
     // delete obsolete states
-    objects.getObjectView('system', 'state', {startkey: 'system.host.' + hostname + '.', endkey: 'system.host.' + hostname + '.\u9999', include_docs: true}, (_err, doc) => {
+    objects.getObjectView('system', 'state', {startkey: hostObjectPrefix + '.', endkey: hostObjectPrefix + '.\u9999', include_docs: true}, (_err, doc) => {
         // identify existing states for deletion, because they are not in the new tasks-list
-        const todelete = doc.rows.filter(out1 => !tasks.some(out2 => out1.id === out2._id));
+        let thishostStates = doc.rows;
+        if (!compactGroupController) {
+            thishostStates = doc.rows.filter(out1 => !out1.id.includes(hostObjectPrefix + '.compactgroup'));
```
Comment (AlCalzone) [inline main.js]: I think `hostObjectPrefix + '.compactgroup'` should be a constant at the top somewhere.

File: main.js [hunk 20]
```diff
@@ -1884,27 +2053,57 @@ function getInstances() {
...
-                logger.warn('host.' + hostname + ' does not start any instances on this host');
+                logger.warn(hostLogPrefix + ' does not start any instances on this host');
             }
         }
 
         initInstances();
     });
 }
 
+function instanceRelevantForThisHost(instance, _ipArr) {
```
Comment (AlCalzone) [inline main.js]: Short comment what this function does pls. For which cases does it return true, when false?

File: main.js [hunk 21]
```diff
@@ -1884,27 +2053,57 @@ function getInstances() {
...
+function checkAndAddInstance(instance, ipArr) {
+    if (!instanceRelevantForThisHost(instance, ipArr)) return false;
+    if (config.system.compact && instance.common.compact) {
+        if (instance.common.runAsCompactMode) {
+            compactProcs[instance.common.compactGroup] = compactProcs[instance.common.compactGroup] || {instances: []};
+        }
+    }
+
+    if (compactGroupController) {
+        logger.debug(hostLogPrefix + ' instance ' + instance._id + ' is relevant for this controller');
```
Comment (AlCalzone) [inline main.js]: How about `managed by` or `belongs to` instead of `relevant for`?

File: main.js [hunk 22]
```diff
@@ -2471,44 +2825,52 @@ function startInstance(id, wakeUp) {
...
                 }
             } else {
-                !wakeUp && procs[id] && logger.warn('host.' + hostname + ' instance ' + instance._id + ' already running with pid ' + procs[id].process.pid);
+                !wakeUp && procs[id] && logger.warn(hostLogPrefix + ' instance ' + instance._id + ' already running with pid ' + procs[id].process.pid);
             }
             break;
 
         case 'schedule':
+            if (compactGroupController) {
+                logger.debug(hostLogPrefix + ' ' + instance._id + ' schedule is not started by compact group controller');
```
Comment (AlCalzone) [inline main.js]: Would it make sense to have a special compact group controller for schedules then?

File: lib/setup.js [hunk 23]
```diff
@@ -2204,12 +2204,13 @@ function dbConnect(onlyCheck, params, callback) {
         params    = null;
     }
     params = params || {};
+
+    const config = JSON.parse(fs.readFileSync(tools.getConfigFileName(), 'utf8'));
```
Comment (AlCalzone) [inline lib/setup.js]: I read this exact line so often. Shouldn't that be a utility function somewhere? E.g. in `tools`

File: lib/setup/setupInstall.js (hunk 1 above)
Comment (Apollon77) [reply lib/setup/setupInstall.js]: fixed

File: lib/states/statesInRedis.js (hunk 2 above)
Comment (Apollon77) [reply lib/states/statesInRedis.js]: Done, changed to socket.io

File: lib/cli/cliCompact.js (hunk 3 above)
Comment (Apollon77) [reply lib/cli/cliCompact.js]: Done

File: lib/cli/cliCompact.js (hunk 4 above)
Comment (Apollon77) [reply lib/cli/cliCompact.js]: yes "restart" is better

File: lib/cli/cliCompact.js (hunk 5 above)
Comment (Apollon77) [reply lib/cli/cliCompact.js]: I changed to "You need to enable compact mode for this host for this setting to take effect!"

File: lib/cli/cliCompact.js (hunk 6 above)
Comment (Apollon77) [reply lib/cli/cliCompact.js]: changed both places again ...

File: main.js (hunk 8 above)
Comment (Apollon77) [reply main.js]: because only the "real" host controller is handling messages (execure commands for this host and such) and so only relevant if "not" a compact group controller

File: main.js (hunk 9 above)
Comment (Apollon77) [reply main.js]: Done

File: main.js (hunk 11 above)
Comment (Apollon77) [reply main.js]: fixed

File: main.js (hunk 12 above)
Comment (Apollon77) [reply main.js]: Ok, they were that way before ... webExtensions and webInstance adapters do not have a process ... also not when disabled

File: main.js (hunk 13 above)
Comment (Apollon77) [reply main.js]: because setting can bié different per host (in multihost env) and status is currently only in "core config json" ... so having it visible as state makes sense for me, also to show in Admin or such

File: main.js (hunk 14 above)
Comment (Apollon77) [reply main.js]: compactGroupProcesses = "number of compact group controllern" (not adapter or instances!)

File: main.js (hunk 16 above)
Comment (Apollon77) [reply main.js]: added: only main host controller needs to check/fix the host assignments from the instances

File: main.js (hunk 17 above)
Comment (Apollon77) [reply main.js]: fixed

File: main.js (hunk 18 above)
Comment (Apollon77) [reply main.js]: i did not touched this line ;-)
In fact process.execArgv is special because these in there are not in regular argv ... so they are added and then the one is removed

File: main.js (hunk 19 above)
Comment (Apollon77) [reply main.js]: no, only one codeplace is using this. in all normal cases the "init()" method will add it to hostObjecPrefix already

File: main.js (hunk 20 above)
Comment (Apollon77) [reply main.js]: done

File: main.js (hunk 21 above)
Comment (Apollon77) [reply main.js]: managed by ...

File: main.js (hunk 22 above)
Comment (Apollon77) [reply main.js]: currently schedules are not handled by compat at all ... this would be more a "feature request "for later :-)

Comment (Apollon77) [reply main.js]: I do not wanted to make it even bigger. In theory oit should be comparable ... but then memory footprint would be bigger because code is loaded also if only running once a day ...

File: lib/setup.js (hunk 23 above)
Comment (Apollon77) [reply lib/setup.js]: With my change in "dbConnect" to return config in the callback as last parameter all other "setup" places could be changed later to use this instead of including by themself. And then only this one place is left ;-)

File: main.js (hunk 10 above)
Comment (Apollon77) [reply main.js]: There was a timing thinigy where the main host controller should behave slighly different. I schanged to 500ms ... will check while final testing - and ad a comment

File: main.js (hunk 19 above)
Comment (AlCalzone) [reply main.js]: I did see that a few times at the bottom. Just search for lowercase compactgroup

File: main.js (hunk 12 above)
Comment (AlCalzone) [reply main.js]: It was there before doesn't mean a comment is not necessary 😉

File: main.js (hunk 19 above)
Comment (Apollon77) [reply main.js]: Yes but should only be in „init“ in : lines directly under each other. I check again.

Comment (AlCalzone) [reply main.js]: And here:
https://github.com/ioBroker/ioBroker.js-controller/blob/428a72c05a9537aacec5ed9ecbe718513c347195/main.js#L973

IMO, using the same magic string twice already warrants a constant.

Comment (Apollon77) [reply main.js]: ahhh so you only mean the "compactgroup-postfix" :-) yes will do
